server/rating_service/rating_service.py

```diff
@@ -0,0 +1,382 @@
+    ) -> Rating:
+        if self._rating_type_ids is None:
+            self._logger.warning(
+                "Tried to fetch player data before initializing service."
+            )
+            raise ServiceNotReadyError("RatingService not yet initialized.")
+
+        rating_type_id = self._rating_type_ids.get(rating_type.value)
+        if rating_type_id is None:
+            raise ValueError(f"Unknown rating type {rating_type}.")
+
+        async with self._db.acquire() as conn:
+            sql = select([rating_table.c.mean, rating_table.c.deviation]).where(
+                and_(
+                    rating_table.c.login_id == player_id,
+                    rating_table.c.leaderboard_id == rating_type_id,
+                )
+            )
+
+            result = await conn.execute(sql)
+            row = await result.fetchone()
+
+        if not row:
+            return await self._get_player_legacy_rating(player_id, rating_type)
+
+        return Rating(row[rating_table.c.mean], row[rating_table.c.deviation])
+
+    async def _get_player_legacy_rating(
+        self, player_id: int, rating_type: RatingType
+    ) -> Rating:
+        if rating_type is RatingType.GLOBAL:
+            table = legacy_global_table
+            sql = select([table.c.mean, table.c.deviation, table.c.numGames]).where(
+                table.c.id == player_id
+            )
+        elif rating_type is RatingType.LADDER_1V1:
+            table = legacy_ladder1v1_table
+            sql = select(
+                [table.c.mean, table.c.deviation, table.c.numGames, table.c.winGames]
+            ).where(table.c.id == player_id)
+        else:
+            raise ValueError(f"Unknown rating type {rating_type}.")
+
+        async with self._db.acquire() as conn:
+
+            result = await conn.execute(sql)
+            row = await result.fetchone()
+
+            if not row:
+                new_rating = await self._create_default_rating(
+                    conn, player_id, rating_type
+                )
+                return new_rating
+
+            if rating_type is RatingType.GLOBAL:
+                # The old `global_rating` table does not have a `winGames` column.
+                # This should be a decent approximation though.
+                won_games = row[table.c.numGames] // 2
+            else:
+                won_games = row[table.c.winGames]
+
+            await conn.execute(
+                "INSERT INTO leaderboard_rating "
+                "(login_id, mean, deviation, total_games, won_games, leaderboard_id) "
+                "VALUES (%s, %s, %s, %s, %s, %s)",
+                player_id,
+                row[table.c.mean],
+                row[table.c.deviation],
+                row[table.c.numGames],
+                won_games,
+                self._rating_type_ids[rating_type.value],
+            )
+
+            return Rating(row[table.c.mean], row[table.c.deviation])
+
+    async def _create_default_rating(
```

Comment:
The only thing I'm worried about with doing the creation here is that the matchmaker for example won't have a rating entry to use for matchmaking until AFTER someone plays a game. I think it makes more sense to do this on first login in `PlayerService`.

Reply:
The `Player` objects will still provide a default rating of `(1500, 500)` that is hardcoded in their initialization (unless I'm missing something) but not written to the database until it changes.
I'll at least make sure that that default value is now chosen according to the config variable.

Reply:
I think I like that all writes to the rating tables are in the same place (the rating service), even if it means there is this ambiguity about default ratings...
Note that the default value written here only exists so that the rating change will be an `UPDATE` instead of an `INSERT` and can therefore also be left out if there is a slick way to "`UPSERT`".

Let me know if you disagree :)

Reply:
We could probably do this for writing rating changes:
https://docs.sqlalchemy.org/en/13/dialects/mysql.html#insert-on-duplicate-key-update-upsert

Reply:
Ah no that doesn't work, since the table doesn't list (`login_id`, `leaderboard_id`) pairs as unique I think.

Reply:
Ok, as long as the player object has the default rating too then I guess that's fine.